### PR TITLE
Add `leaves()` to make it simple to collect all parameters in a model

### PR DESCRIPTION
**Pull request by staticfloat**
Useful for implementing things like L2 regularization, typically
requires a filtering afterwards such as collecting only
`AbstractArrays`, etc...

### CODE REVIEW

File: src/treelike.jl [hunk 1]
```diff
@@ -9,10 +9,26 @@ children(x::NamedTuple) = x
...
+  end
+  # Return the `params` collection in the end
+  return params
+end
+
 function treelike(m::Module, T, fs = fieldnames(T))
   @eval m begin
     Flux.children(x::$T) = ($([:(x.$f) for f in fs]...),)
-    Flux.mapchildren(f, x::$T) = $T(f.($children(x))...)
+    Flux.mapchildren(f, x::$T) = $T(map(f, children(x))...)
```
Comment (staticfloat) [inline src/treelike.jl]: I changed this to `map` just to be consistent with the `mapchildren()` definition above on line 10

**Comment by MikeInnes** [comment]
Can you talk a bit about how this compares with `params`?

**Comment by staticfloat** [comment]
Oh, I didn't really know about `Flux.params()`, but it doesn't seem to work the same way:

```julia
using Flux, Metalhead
m = Metalhead.VGG19()

@show length(params(m))
@show length(leaves(m))
```

```
length(params(m)) = 0
length(leaves(m)) = 162
```

**Comment by MikeInnes** [comment]
Right now `params` filters out `TrackedArray`, but we'll need to remove that restriction anyway so that things work with Zygote. Removing that should get you the behaviour you want from `leaves` as well, I think.

**Comment by staticfloat** [comment]
Cool, then I think there's nothing to see here.

**Pull request closed without merging**